src/charm.py

```diff
@@ -12,92 +12,292 @@
+            self.unit.status = BlockedStatus("Missing relations: ceph")
         else:
-            event.set_results({"fortune": "A bug in the code is worth two in the documentation."})
+            self.unit.status = ActiveStatus()
+
+    @needs_leader
+    def create_ceph_resources(self, resources: List[Dict]) -> None:  # pylint: disable=R0201
+        """Use kubernetes api to create resources like Pods, Secrets, etc.
+
+        :param resources: list of dictionaries describing resources
+        :return: None
+        """
+        config.load_kube_config()
+        k8s_api = client.ApiClient()
+
+        for resource in resources:
+            utils.create_from_dict(k8s_api, resource)
+
+    def render_resource_definitions(self) -> List[Dict]:
+        """Render resource definitions from templates in self.RESOURCE_TEMPLATES."""
+        env = Environment(loader=FileSystemLoader(self.template_dir))
+
+        resources = [
+            env.get_template(template).render(self._stored.ceph_data)
+            for template in self.RESOURCE_TEMPLATES
+        ]
+
+        resource_dicts = [yaml.safe_load_all(res) for res in resources]
+        return list(itertools.chain.from_iterable(resource_dicts))
+
+    def render_storage_definitions(self) -> List[Dict]:
+        """Render StorageClass definitions for supported filesystem types."""
+        env = Environment(loader=FileSystemLoader(self.template_dir))
+        default_storage = self._stored.default_storage_class
+        storage_classes = []
+
+        ext4_ctx = self.copy_stored_dict(self._stored.ceph_data)
+        ext4_ctx["default"] = default_storage == self.EXT4_STORAGE
+        ext4_ctx["pool_name"] = "ext4-pool"
+        ext4_ctx["fs_type"] = "ext4"
+        ext4_ctx["sc_name"] = "ceph-ext4"
+        resource = env.get_template(self.STORAGE_CLASS_TEMPLATE).render(ext4_ctx)
+        storage_classes.append(yaml.safe_load(resource))
+
+        xfs_ctx = self.copy_stored_dict(self._stored.ceph_data)
+        xfs_ctx["default"] = default_storage == self.XFS_STORAGE
+        xfs_ctx["pool_name"] = "xfs-pool"
+        xfs_ctx["fs_type"] = "xfs"
+        xfs_ctx["sc_name"] = "ceph-xfs"
+        resource = env.get_template(self.STORAGE_CLASS_TEMPLATE).render(xfs_ctx)
+        storage_classes.append(yaml.safe_load(resource))
+
+        return storage_classes
+
+    def render_all_resource_definitions(self) -> List[Dict]:
+        """Render all resources required for ceph-csi."""
+        return self.render_resource_definitions() + self.render_storage_definitions()
+
+    @needs_leader
+    def update_storage_classes(self) -> None:
+        """Re-render templates and update StorageClass resources in k8s cluster."""
+        for resource in self.resources:
+            if isinstance(resource, StorageClass):
+                resource.remove()
+        storage_classes = self.render_storage_definitions()
+        self.create_ceph_resources(storage_classes)
```

Comment:
I wonder about adding 3 lines of inline comments to separate helper functions and hooks-related functions. This is completely nitpick, since everything ends in the same class definition.

Reply:
Sorry, I don't understand this one.

Reply:
This is more of a visual OCD. The suggestion goes along the lines of adding:

```
#
# Helper functions
#
def ...

#
# Observed functions (triggered by hooks)
#
def _on_config_changed(...):
```

Ideally, the helper functions could be part of a lib which accepts arguments with class attributes (so the calls to self.helper_method wouldn't be needed).

Again, please ignore if you don't find this helpful :)